### DISCUSSION TITLE
Interesting library but a few remarks

### DISCUSSION THREAD

**Discussion by Nicolas4449** [ta4j/ta4j#917] [General]
Hello,

Sorry if everything is not very understandable, I use a translator.

TA4j is very interesting, well done. I have been trying to develop an automatic trading system for a few weeks.
I will maybe migrate a part to TA4J (for the indicators) because it is more complete than my own engine, but for the rest, looking at the source code, I have some remarks, questions or ideas before using Strategy and BackTesting:
   - there is a single entry and exit trade on positions ?. Is it possible to manage positions with several entry and/or exit trades (ex: pyramid orders, exit in two stages or partial execution of orders)?
   - what are the Backtesting performances? My system takes an average of 2 seconds for a test of two years of data with a 15 minute period and a medium term trading strategy based on 4 to 6 indicators. This is important to be able to mass test different parameters.
   - in the rules, I did not really understand if it is possible: is it possible to set up strategies in two stages?. For example, if the RSI exceeds a value, then we start monitoring the other rules without taking the value of the RSI into account. So one or more rules of the strategy are satisfied at t, and the other rules are satisfied at t+x?
   - it seems to be missing the regression oscillator indicator? If I use ta4j I might add it.
   
   - FYI, for the unit back tests, I try to export the data from my tests to then consult them with TradingVue.js. For mass, long-term back-testing, I'm thinking of integrating a system like simulated annealing (or another) to find the best parameters.
   - is it planned to add Interfaces to connect to Trading platforms?

Thanks for the work and for sharing.

numomeil

**Comment by nimo23** [comment]
Thanks for sharing your ideas:

> there is a single entry and exit trade on positions ?. Is it possible to manage positions with several entry and/or exit trades (ex: pyramid orders, exit in two stages or partial execution of orders)?

See https://github.com/ta4j/ta4j/issues/755 and share your ideas with us.

>what are the Backtesting performances? My system takes an average of 2 seconds for a test of two years of data with a 15 minute period and a medium term trading strategy based on 4 to 6 indicators. This is important to be able to mass test different parameters.

The performances depend on many factors (e.g. using DecimalNum or DoubleNum, etc.) However, it also depends on the kind of backtesting. What are you using for a kind of backtesting methodology?

>in the rules, I did not really understand if it is possible: is it possible to set up strategies in two stages?. For example, if the RSI exceeds a value, then we start monitoring the other rules without taking the value of the RSI into account. So one or more rules of the strategy are satisfied at t, and the other rules are satisfied at t+x?

Yes, this is possible. See https://github.com/ta4j/ta4j/blob/develop/ta4j-core/src/main/java/org/ta4j/core/rules/ChainRule.java

>it seems to be missing the regression oscillator indicator? If I use ta4j I might add it.

Please add a PR. That would be cool!

>is it planned to add Interfaces to connect to Trading platforms?

No, this would not be a good idea. It would lower the degree of specialization and maintainability of this project as many things can change here at a rate (connection and API staff from brokers or 3rd party libs) that this project would not be able to keep up with. If you want broker integration then use something like xChange, etc. Ta4j has very little 3rd party dependency (of fast changing APIs) and this should always be the case. With this, you can be sure, that this lib works in the future even when there are no maintainers anymore:)
